### PR TITLE
Leetcode optimizations

### PR DESCRIPTION
**Pull request by vhavlena**
Optimizations for Leetcode

### CODE REVIEW

File: src/smt/theory_str_noodler/var_union_find.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    static bool ec_are_equal(const BasicTermEqiv& ec, const BasicTerm& t1, const BasicTerm& t2) {
+        for(const auto& st: ec) {
+            if(st.find(t1) != st.end() && st.find(t2) != st.end()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    class var_union_find {
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/var_union_find.h:39]: Some comments about what this class is for would be nice

File: src/smt/theory_str_noodler/decision_procedure.h [hunk 2]
```diff
@@ -333,11 +335,13 @@ namespace smt::noodler {
          * @param m Z3 AST manager
          * @param m_util_s Z3 string manager
          * @param m_util_a Z3 arithmetic manager
+         * @param len_eq_vars Equivalence clsss holding variables with the same length
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/decision_procedure.h:338]:
```suggestion
         * @param len_eq_vars Equivalence class holding variables with the same length
```

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 3]
```diff
@@ -1035,6 +1084,78 @@ namespace smt::noodler {
...
+        }
+    }
+
+    /**
+     * @brief Check if two languages of two terms have the same length.
+     * 
+     * @param ec Equivalence class containing length-equivalent variables.
+     * @param t1 Term1
+     * @param t2 Term2
+     * @return Equial length -> true
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/formula_preprocess.cpp:1146]:
```suggestion
     * @return Equal length -> true
```

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 4]
```diff
@@ -787,7 +787,16 @@ namespace smt::noodler {
             return FC_DONE;
         }
 
-        DecisionProcedure dec_proc = DecisionProcedure{ instance, aut_assignment, init_length_sensitive_vars, m, m_util_s, m_util_a, m_params };
+        // for(const auto& t : this->var_eqs.get_equivalence_bt()) {
+        //     for (const auto& s : t) {
+        //         std::cout << s.to_string() << " ";
+        //     }
+        //     std::cout << std::endl;
+        // }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:795]: Maybe put this in STRACE? Otherwise delete it.

**Review by jurajsic — APPROVED** [review]
I think it is fine, but I can't say I fully understand what is happening in some preprocessing steps.

**Review by Adda0 — APPROVED** [review]
I do not see anything critical. Other than the mentioned details, I approve the PR. If the benchmarks pass, I am OK with merging the PR.

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 5]
```diff
@@ -436,7 +468,12 @@ namespace smt::noodler {
                     std::vector<BasicTerm>({val2.term})}));
                 return true;
             }
-        } // TODO: allow to generate X = eps
+        } else if(diff.first.size() == 1) {
+            return propagate_regular_eqs(diff.first, diff.second, new_pred);
+        } else if(diff.second.size() == 1) {
+            return propagate_regular_eqs(diff.second, diff.first, new_pred);
+        }
+        /// TODO: allow to generate X = eps
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:476]: Yes, still relevant

File: src/smt/theory_str_noodler/aut_assignment.h [hunk 6]
```diff
@@ -121,6 +121,23 @@ namespace smt::noodler {
...
+        /**
+         * @brief Check if the language of the basic term has a fixed length
+         * 
+         * @param t BasicTerm
+         * @param n[out] fixed length
+         * @return True->fixed length
+         */
+        bool fixed_length(const BasicTerm& t, int& n) const {
+            auto lengths = Mata::Strings::get_word_lengths(*this->at(t));
+            if(lengths.size() > 1) {
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/aut_assignment.h]: Fixed

File: src/smt/theory_str_noodler/var_union_find.h (hunk 1 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/var_union_find.h:44]: I added some comments

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 7]
```diff
@@ -698,6 +702,14 @@ namespace smt::noodler {
         this->prep_handler.remove_trivial();
         this->prep_handler.reduce_regular_sequence(3);
         this->prep_handler.remove_regular();
+
+        // the following should help with Leetcode
+        /// TODO: should be simplyfied? So many preprocessing steps now
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.cpp:707]: Yes, we need to think about it at some point

**Comment by vhavlena** [comment]
If you don't have another comments I will merge it

**Comment by Adda0** [comment]
I approve.

**Comment by jurajsic** [comment]
We can merge this

**Pull request merged**